cmd/panasas/config/client.go

```diff
@@ -0,0 +1,389 @@
+package config
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+const panasasHTTPS3PrefixHeader string = "x-panasas-s3-prefix"
+const panasasHTTPS3MetadataHeader string = "x-panasas-s3-metadata"
+const panasasHTTPS3ConfigRevisionHeader string = "x-panasas-s3-config-revision"
+
+// ErrNotFound informs that the requested object has not been found by the
+// config agent.
+var ErrNotFound = fmt.Errorf("Not found")
+
+// ErrNoRevisionYet informs that no successful modifying operation has been
+// performed yet and consequently there is not cached config revision.
+var ErrNoRevisionYet = fmt.Errorf("Config revision not retrieved yet")
+
+// ErrUnexpectedHTTPStatus informs about HTTP status value outside the expected
+// range.
+type ErrUnexpectedHTTPStatus uint
+
+// ErrUnexpectedContentType informs about unexpected HTTP response content type
+type ErrUnexpectedContentType string
+
+func (e ErrUnexpectedHTTPStatus) Error() string {
+	return fmt.Sprintf("Unexpected HTTP status: %v", uint(e))
+}
+
+func (e ErrUnexpectedContentType) Error() string {
+	return fmt.Sprintf("Unexpected HTTP content type: %q", string(e))
+}
+
+// PanasasConfigClient represents a Panasas config agent client
+type PanasasConfigClient struct {
+	agentHost  string
+	agentPort  uint
+	httpClient *http.Client
+
+	configRevision *string
+}
+
+// NewPanasasConfigClient returns a configured PanasasConfigClient
+func NewPanasasConfigClient(host string, port uint) *PanasasConfigClient {
+	client := PanasasConfigClient{
+		agentHost:  host,
+		agentPort:  port,
+		httpClient: &http.Client{},
+	}
+	return &client
+}
+
+func (c *PanasasConfigClient) getPanasasConfigStoreURL(base string, elem ...string) (*url.URL, error) {
+	var path string = base
+	var e error
+
+	for _, eachElem := range elem {
+		path, e = url.JoinPath(base, url.PathEscape(eachElem))
+		if e != nil {
+			return nil, e
+		}
+	}
+
+	return url.Parse(
+		fmt.Sprintf("http://%v:%v%v", c.agentHost, c.agentPort, path),
```

Comment:
Will we always use the HTTP here? Maybe it makes sense to put it info in `PanasasConfigClient` structure?